rf: local variable rename panic

Attempting to rename a variable in a function results in a panic in `rewriteDefn`. This occurs for the documentation example of `mv F.who F.greeter`, or as a test case:

```
mv F.x F.y
-- x.go --
package m

func F() {
        x := 123
        _ = x
}
-- stdout --
diff old/x.go new/x.go
--- old/x.go
+++ new/x.go
@@ -1,6 +1,6 @@
 package m

 func F() {
-       x := 123
-       _ = x
+       y := 123
+       _ = y
 }
```

```
$ go test
panic executing: mv F.x F.y
--- FAIL: TestRun (13.06s)
    --- FAIL: TestRun/mv_localvar.txt (0.05s)
        rf_test.go:85: testdata/mv_localvar.txt
panic: [*ast.Ident *ast.AssignStmt *ast.BlockStmt *ast.FuncDecl *ast.File] [recovered]
        panic: [*ast.Ident *ast.AssignStmt *ast.BlockStmt *ast.FuncDecl *ast.File] [recovered]
        panic: [*ast.Ident *ast.AssignStmt *ast.BlockStmt *ast.FuncDecl *ast.File]

goroutine 510 [running]:
testing.tRunner.func1.2(0x82a440, 0xc00577c190)
        /usr/lib/google-golang/src/testing/testing.go:1144 +0x332
testing.tRunner.func1(0xc00966c600)
        /usr/lib/google-golang/src/testing/testing.go:1147 +0x4b6
panic(0x82a440, 0xc00577c190)
        /usr/lib/google-golang/src/runtime/panic.go:965 +0x1b9
rsc.io/rf.run.func1(0xc00541dd90)
        /usr/local/google/home/mpratt/src/rf/rf.go:76 +0xe7
panic(0x82a440, 0xc00577c190)
        /usr/lib/google-golang/src/runtime/panic.go:965 +0x1b9
rsc.io/rf.rewriteDefn(0xc005778000, 0xc005776340, 0xc00577a07c, 0x1)
        /usr/local/google/home/mpratt/src/rf/mv.go:284 +0x78c
rsc.io/rf.cmdMv(0xc005778000, 0xc004bbf3e3, 0x7)
        /usr/local/google/home/mpratt/src/rf/mv.go:190 +0x123f
rsc.io/rf.run(0xc00305fc20, 0xc004bbf3d0, 0xb, 0x0, 0x0)
        /usr/local/google/home/mpratt/src/rf/rf.go:133 +0x30f
rsc.io/rf.TestRun.func1(0xc00966c600)
        /usr/local/google/home/mpratt/src/rf/rf_test.go:121 +0x48a
testing.tRunner(0xc00966c600, 0xc004bc5880)
        /usr/lib/google-golang/src/testing/testing.go:1194 +0xef
created by testing.(*T).Run
        /usr/lib/google-golang/src/testing/testing.go:1239 +0x2b3
exit status 2
FAIL    rsc.io/rf       13.093s
```

(Note I hand-edited the diff to what I expect the output to be, but it might be slightly off).

--- a/mv.go
+++ b/mv.go
@@ -181,7 +181,7 @@ func cmdMv(snap *refactor.Snapshot, args string) {
 		return
 	}
 
-	// Rename of struct field or method.
+	// Rename of local variable, struct field, or method.
 	if old.Outer != nil && newOuter != nil && old.Outer.Obj == newOuter.Obj {
 		if newItem.Kind != refactor.ItemNotFound { // TODO
 			snap.ErrorAt(newItem.Obj.Pos(), "already have %s", newPath)
@@ -282,7 +282,7 @@ func rewriteDefn(snap *refactor.Snapshot, old *refactor.Item, new string) {
 	switch outer.(type) {
 	default:
 		panic(refactor.StackTypes(stack))
-	case *ast.FuncDecl, *ast.Field:
+	case *ast.FuncDecl, *ast.Field, *ast.AssignStmt:
 		// ok
 	case *ast.ValueSpec, *ast.TypeSpec:
 		decl := stack[2].(*ast.GenDecl)
@@ -306,7 +306,7 @@ func rewriteUses(snap *refactor.Snapshot, old *refactor.Item, new string, checkP
 			id, ok := stack[0].(*ast.Ident)
 			if !ok || pkg.TypesInfo.Uses[id] != old.Obj {
 				if len(stack) > 2 {
-					if _, ok := stack[1].(*ast.AssignStmt); ok && pkg.TypesInfo.Defs[id] == old.Obj {
+					if as, ok := stack[1].(*ast.AssignStmt); ok && pkg.TypesInfo.Defs[id] == old.Obj && as.Tok != token.DEFINE {
 						goto OK
 					}
 				}
@@ -372,7 +372,10 @@ func removeDecl(snap *refactor.Snapshot, old *refactor.Item) string {
 			return ""
 		}
 		docPos, _ := nodeRange(snap, stack[1])
-		snap.DeleteAt(as.TokPos, as.TokPos+1) // delete colon
+		// Delete colon and update as so other users can tell this is
+		// no longer a declaration.
+		snap.DeleteAt(as.TokPos, as.TokPos+1)
+		as.Tok = token.ASSIGN
 		return string(snap.Text(docPos, as.Pos()))
 	}
 

--- a/testdata/mv_localshortvar.txt
+++ b/testdata/mv_localshortvar.txt
@@ -1,0 +1,21 @@
+mv F.x F.y
+-- x.go --
+package m
+
+func F() {
+	x := 123
+	_ = x
+}
+-- stdout --
+diff old/x.go new/x.go
+--- old/x.go
++++ new/x.go
+@@ -1,6 +1,6 @@
+ package m
+
+ func F() {
+-	x := 123
+-	_ = x
++	y := 123
++	_ = y
+ }

--- a/testdata/mv_localvar.txt
+++ b/testdata/mv_localvar.txt
@@ -1,0 +1,21 @@
+mv F.x F.y
+-- x.go --
+package m
+
+func F() {
+	var x int = 123
+	_ = x
+}
+-- stdout --
+diff old/x.go new/x.go
+--- old/x.go
++++ new/x.go
+@@ -1,6 +1,6 @@
+ package m
+
+ func F() {
+-	var x int = 123
+-	_ = x
++	var y int = 123
++	_ = y
+ }

